### PR TITLE
[Merged by Bors] - Refactor merging and validation of the configuration

### PR DESCRIPTION
**Pull request by siegfriedweber**
# Description

Merging and validation of the configuration refactored

Prerequisite for #364 

## Integration tests

[Build Status](https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/druid-operator-it-custom/81/)

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Review by maltesander — APPROVED** [review]
I like it. Just really minor things not even from this PR :D sorry for that. LGTM otherwise.

File: rust/crd/src/resource.rs [hunk 1]
```diff
@@ -51,34 +36,12 @@ pub enum Error {
...
-            RoleResourceFragment::DruidFragment(fragment) => Ok(RoleResource::Druid(
-                fragment::validate(fragment).with_context(|_| ResourceValidationSnafu)?,
-            )),
-            RoleResourceFragment::HistoricalFragment(fragment) => Ok(RoleResource::Historical(
-                fragment::validate(fragment).with_context(|_| ResourceValidationSnafu)?,
-            )),
-        }
-    }
-}
-
```
[the remark this answers is not in the document]
Comment (siegfriedweber) [reply rust/crd/src/resource.rs:81]: Fixed in 90bb2e6543a3dfe7172f2155352f935f9d4a2a30

File: rust/operator-binary/src/druid_controller.rs [hunk 2]
```diff
@@ -311,6 +302,8 @@ pub async fn reconcile_druid(druid: Arc<DruidCluster>, ctx: Arc<Ctx>) -> Result<
     )
     .context(CreateClusterResourcesSnafu)?;
 
+    let merged_config = druid.merged_config().context(FailedToResolveConfigSnafu)?;
+
```
[the remark this answers is not in the document]
Comment (siegfriedweber) [reply rust/operator-binary/src/druid_controller.rs:304]: Fixed in 90bb2e6543a3dfe7172f2155352f935f9d4a2a30

File: rust/operator-binary/src/druid_controller.rs [hunk 3]
```diff
@@ -161,17 +159,10 @@ pub enum Error {
...
-    InvalidJavaHeapConfig,
-    #[snafu(display("failed to convert java heap config to unit [{unit}]"))]
-    FailedToConvertJavaHeap {
-        source: stackable_operator::error::Error,
-        unit: String,
-    },
+    #[snafu(display("failed to resolve and merge config for role and role group"))]
+    FailedToResolveConfig { source: stackable_druid_crd::Error },
+    #[snafu(display("invalid configuration"))]
+    InvalidConfiguration { source: stackable_druid_crd::Error },
```
[the remark this answers is not in the document]
Comment (siegfriedweber) [reply rust/operator-binary/src/druid_controller.rs:165]: This error variant is used in line 540:

```rust
                let runtime_properties =
                    stackable_operator::product_config::writer::to_java_properties_string(
                        conf.iter(),
                    )
                    .context(PropertiesWriteSnafu)?;
```

Snafu removes the Error suffix.

**Review by maltesander — APPROVED** [review]
LGTM.

**Comment by siegfriedweber** [comment]
bors merge

**Pull request closed without merging**